objects/src/block/mod.rs

```diff
@@ -1,6 +1,71 @@
+use alloc::vec::Vec;
+
 use super::{Digest, Felt, Hasher, ZERO};
 
 mod header;
 pub use header::BlockHeader;
 mod note_tree;
 pub use note_tree::BlockNoteTree;
+
+use crate::{
+    notes::Nullifier,
+    transaction::{AccountUpdateData, OutputNote},
+};
+
+pub type NoteBatch = Vec<OutputNote>;
+
+/// A block in the block chain.
+#[derive(Debug, Clone)]
+pub struct Block {
+    /// Block header.
+    header: BlockHeader,
+
+    /// Account updates for the block.
+    updated_accounts: Vec<AccountUpdateData>,
+
+    /// Note batches created in transactions in the block.
+    created_notes: Vec<NoteBatch>,
+
+    /// Nullifiers produced in transactions in the block.
+    produced_nullifiers: Vec<Nullifier>,
+    // TODO:
+    // - full states for created public notes
+    // - zk proof
+}
+
+impl Block {
+    /// Creates a new block.
+    pub const fn new(
+        header: BlockHeader,
+        updated_accounts: Vec<AccountUpdateData>,
+        created_notes: Vec<NoteBatch>,
+        produced_nullifiers: Vec<Nullifier>,
+    ) -> Self {
+        Self {
+            header,
+            updated_accounts,
+            created_notes,
+            produced_nullifiers,
+        }
+    }
+
+    /// Returns the block header.
+    pub fn header(&self) -> BlockHeader {
+        self.header
+    }
+
+    /// Returns the account updates.
+    pub fn updated_accounts(&self) -> &Vec<AccountUpdateData> {
+        &self.updated_accounts
+    }
+
+    /// Returns the note batches.
+    pub fn created_notes(&self) -> &Vec<NoteBatch> {
+        &self.created_notes
+    }
+
+    /// Returns the nullifiers.
+    pub fn produced_nullifiers(&self) -> &Vec<Nullifier> {
+        &self.produced_nullifiers
+    }
```

Comment:
nit: instead of returning `&Vec` references, I'd probably return `&[]` references.

Reply:
As a general principle I agree with the logic, but in case of `&Vec` vs. `&[]` I'm not sure it makes a difference. Specifically, there are very few methods which `&Vec` has which `&[]` doesn't (the only one that comes to mind is `Vec::capacity()` - everything else requires a mutable reference). So, practically, we don't gain any additional functionality be returning `&Vec`.

The situation would be different if we were returning `&mut Vec` - but in case of immutable `&Vec`, I think we should replace it with just `&[]` as it is cleaner and doesn't reduce functionality.

Reply:
I've searched for usage of `&Vec` and now I think that you're right, it's better to use `&[]` everywhere. I will fix this.